Fix merging of property metadata

TimvdLippe opened the pull request:
Before `this._propertyInfo` was unused. It took the first occurrence of `this.properties[property]` and returned only this metadata. As described in #2242, all metadata of a super behavior is ignored. Additionally, inconsistencies regarding types were discovered in #2493.

This pull-request corrects the construction `this._propertyInfo` and let `this.getPropertyInfo` utilize this object.

As a result, a performance increase should be observed, as the object is only constructed at element create time. It is therefore not needed to traverse the properties every time `this.getPropertyInfo` is called.

Fixes #2242
Fixes #2493

kevinpschaaf commented:
It looks like this probably has a couple of issues, but @sorvell will take a closer look:
* The new code grabs the first property descriptor object off a behavior/prototype and modifies it; if that comes from a behavior it will have modified it for all users of the behavior
* The change seems to affect "first valued" approach to `type` and `readOnly`

kevinpschaaf commented:
FYI, we didn't make this kinda obvious change when we introduced `_propertyInfo` (which was added purely as an implementation-detail related to performance for areas that needed specific property info in hot paths) out of concern for breaking the existing API of `getPropertyInfo`, so this PR will require us to re-evaluate that concern.

TimvdLippe commented:
@kevinpschaaf I have fixed the first point and added a testcase for it. While it does have a different behavior, I do think the old behavior was inherently broken and the new behavior is what a Polymer user would expect.

The first valued approach would result in different behavior when an element inherits a behavior with `readOnly` true and explicitly sets `readOnly` to false. I am not aware of any use case for this behavior.
A conflict in types is in my eyes something the user should be aware of. This is what a user would expect when using inheritance in OOP.

sorvell commented on src/micro/properties.html:
```diff
@@ -159,17 +132,17 @@
-              type: typeof(s) === 'function' ? s : s.type,
-              readOnly: s.readOnly,
-              attribute: Polymer.CaseMap.camelToDashCase(i)
+            target[i] = {};
```
We can probably use `chainObject` here instead of copying.

TimvdLippe commented:
@sorvell Replaced the property copying with `chainObject`

sorvell commented:
I think it makes sense to block this PR on the resolution of #3418

TimvdLippe commented:
@sorvell is this eligible for merging for 2.0 or shall I close this PR?

TimvdLippe commented:
I hope this is addressed already in 2.0, but given the inactivity on this PR I am going to close it.
